### ISSUE TITLE
[API Proposal]: Simple one-shot hashing methods that produce the relevant type

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#76279]
### Background and motivation

System.IO.Hashing contains types for Crc32, Crc64, XxHash32, and XxHash64, and XxHash3 is being added.  These types all have one-shot static methods for performing hashes over `ReadOnlySpan<byte>`, but all of these methods either allocate a resulting `byte[]` or write the result into a `Span<byte>`.  However, the natural data types expecting in many use cases is just an int/uint for 32-bit or long/ulong for 64-bit hashes.  We should add helpers for this.

### API Proposal

```diff
namespace System.IO.Hashing;

public sealed class Crc32 : NonCryptographicHashAlgorithm
{
+   public static int HashToInt32(ReadOnlySpan<byte> source);
+   public int GetCurrentHashAsInt32();
    ...
}

public sealed class XxHash32 : NonCryptographicHashAlgorithm
{
+   public static int HashToInt32(ReadOnlySpan<byte> source, int seed = 0);
+   public int GetCurrentHashAsInt32();
    ...
}

public sealed class Crc64 : NonCryptographicHashAlgorithm
{
+   public static long HashToInt64(ReadOnlySpan<byte> source);
+   public long GetCurrentHashAsInt64();
    ...
}

public sealed class XxHash64 : NonCryptographicHashAlgorithm
{
+   public static long HashToInt64(ReadOnlySpan<byte> source, long seed = 0);
+   public long GetCurrentHashAsInt64();
    ...
}

public sealed class XxHash3 : NonCryptographicHashAlgorithm
{
+   public static long HashToInt64(ReadOnlySpan<byte> source, long seed = 0);
+   public long GetCurrentHashAsInt64();
    ...
}

public sealed class XxHash128 : NonCryptographicHashAlgorithm // assuming https://github.com/dotnet/runtime/issues/77885
{
+   public static Int128 HashToInt128(ReadOnlySpan<byte> source, long seed = 0);
+   public Int128 GetCurrentHashAsInt128();
    ...
}
```

- All of these implementations already compute the relevant int/long and then write it to either a newly-allocated array or the destination span.  We'd just return it instead from the new method.
- The existing methods write the value to the destination as big-endian.  The int/long-returning methods would simply return the value in machine-endianness.
- Should we use unsigned types instead?  All of the algorithms internally operate in unsigned land.

### API Usage

```csharp
ReadOnlySpan<byte> data = ...;
int hash = XxHash32.HashToInt32(data);
```

**Comment by LeaFrock** [comment]
Hah, that's what I want!

But so sad that I have to wait for .NET 8 😭

**Comment by LeaFrock** [comment]
I have a very small question. Let's take `XxHash64` for example: why is the API `long HashToInt64` rather than `ulong HashToUInt64`? 

Is it just a convention? I have this question because I see many underlying calculations use unsigned number in fact, such as `String.GetHashCode()`,

```csharp
Marvin.ComputeHash32(ref Unsafe.As<char, byte>(ref _firstChar), (uint)_stringLength * 2 /* in bytes, not chars */, (uint)seed, (uint)(seed >> 32));
```

**Comment by bartonjs** [comment]
@stephentoub I've removed ready-for-review because I want @GrabYourPitchforks to acknowledge the existence of the proposal (and perhaps even express opinions) before it comes up in a review meeting.

**Comment by stephentoub** [comment]
Ok, but I think it's important we add this.  Literally every hashing API I've found that others expose just surface the uint/int/ulong/long.  That our APIs force you to either allocate an array or stackalloc some space into which to store the result, figure out that it's been written as a big-endian value, and then convert it back to being a number is a pit of failure.

**Comment by GrabYourPitchforks** [comment]
The only strong opinion I have is that the endianness should be explicit: include it as part of the API name, or force the caller to include it as an enum parameter. All of these digest algorithms are intended to generate __stable__ output across machines. It's an API pit of failure if we make the "simple" API lead developers unwittingly down a path which violates this fundamental property.

**Comment by stephentoub** [comment]
> All of these digest algorithms are intended to generate stable output across machines.

From my perspective, using machine endianness for the hash code does that implicitly.  The API returns an Int32 and you get back 12345 regardless of whether the machine is little or big endian.  Forcing the method into a specific endianness is what violates this, as it means you get 12345 on some machines and 959447040 on others.

**Comment by GrabYourPitchforks** [comment]
> None of the static Hash methods state the order of the bytes that were written out, and the derived types differ as to whether they use little or big endian to do so, both for the statics and for the instance members.

The output of any digest algorithm is raw bytes. There's no concept of "endianness" unless you attempt to interpret those bytes as an integral value. For instance, we don't say that the digest output of SHA256 is big-endian, little-endian, or something else. It's just raw bytes. You're proposing here an API which performs its own opinionated interpretation of the digest output.

Looking through the implementations of all the `[NonCryptographic]HashAlgorithm`-derived classes, I don't immediately see anything which is ambiguous with regard to endianness. All the implementations I'm seeing follow the "everything is raw bytes" pattern. What am I missing?

**Comment by stephentoub** [comment]
> The output of any digest algorithm is raw bytes

Every other example I can find of such hash functions, both in raw libraries we consume (e.g. zlib), in exemplar implementations on which ours are based (e.g. xxhash), in other frameworks, in other .NET implementations of these same hash functions, and even elsewhere in .NET (e.g. BitOperations.Crc32C) operate in terms of int/long, not in terms of bytes.  Even the internal workhorse functions of all of these algorithms are operating on numbers, and then we take a final step of serializing those numbers into bytes.

> What am I missing?

Almost all of the examples I've found are folks who want the values as integers, and thus are having to reconstruct that from the bytes we spent energy producing, e.g.
https://github.com/dotnet/orleans/blob/cf4423ea4d75b4ab8ecf071968a2a4bfd463646d/src/Orleans.Serialization/TypeSystem/TypeCodec.cs#L210
https://github.com/LeaFrock/BloomFilterSharp/blob/662b0475b3acab4d659a1b298dba4423b53399c3/src/BloomFilterSharp.Hashing/XxHash64Function.cs#L11-L12
https://github.com/avafloww/Thaliak/blob/01eb1e0668b399a70c580f2be1c1480a9699ccda/Thaliak.Common.Database/Models/XivRepository.cs#L26-L28
and in some cases doing so in a way that doesn't pay attention to endianness and thus will violate the "same output" concern _because_ we focused on outputting bytes.

**Comment by GrabYourPitchforks** [comment]
> Even the internal workhorse functions of all of these algorithms are operating on numbers, and then we take a final step of serializing those numbers into bytes.

The public API surface shouldn't be concerned with internal implementation details. The API surface should only ever be concerned with the caller-observable output.

> We have these docs for System.IO.Hashing.Crc32, for example

Because the CRC32 spec is deeply weird and isn't written as a typical digest algorithm spec, so callers need to know "this is the version of CRC32 you're calling." The paragraph immediately under the one you quote in the docs even says as such. It's unfortunately not a good model to base a usable API off of. :(

To the samples that you gave, I think they actually provide a strong argument for why any opinionated APIs we add _should_ explicitly state the endianness.

- Orleans: They want the output in little endian.
- BloomFilterSharp: Their call site appears to be buggy, since they're serializing / persisting a machine-endian value.
- Thaliak: They want the output in big endian.

**Comment by stephentoub** [comment]
> - Orleans: They want the output in little endian.

No, they want a number that's going to be the same regardless of architecture.

> BloomFilterSharp: Their call site appears to be buggy, since they're serializing / persisting a machine-endian value.

That's my point: because we've forced them to go through bytes, we've forced them to consider endianness, and they're not... we've created a pit of failure.  An API that returns machine endianness for the number will always produce the same hash value for the same input bytes, regardless of current machine endianness.

**Comment by LeaFrock** [comment]
> BloomFilterSharp: Their call site appears to be buggy, since they're serializing / persisting a machine-endian value

I never consider about it because I'm just developing a filter using local memory, so the machine endianness isn't a problem. But you're right as the hash function could be called on different machines on a distributed scenario.

Now the key point of divergence is whether the top-level developers `want a number that's going to be the same regardless of architecture`. Both of your opinions sound reasonable depending on the concrete scenarios.

>  An API that returns machine endianness for the number will always produce the same hash value for the same input bytes, regardless of current machine endianness.

I support it for most requirements. But considering for a situation that people want to appoint the concrete endian for compatible with interactions between existing systems (which may be based on other languages) or machines, it'd be better to retain some flexibility. *(You may argue that people should use `byte[]` API but you can't guarantee that)*

I think this is more appropriate for the analogy of `System.StringComparison`. Could it be possible that we add a similar enum for the proposed API?

```csharp

public static long HashToInt64(ReadOnlySpan<byte> source, long seed = 0, EndianOption endianness = default);

public enum EndianOption // don’t care about the name
{
       Default = 0, // As stephentoub says, always return the same value regardless of machine
       Current, // Obey the current machine endian
       Big,
       Little,
}
```

Or even keep a `HashToOptions` for potential extensibility,

```csharp

public static long HashToInt64(ReadOnlySpan<byte> source, long seed = 0, HashToOptions? option = null);

public struct HashToOptions
{
      public EndianOption EndianOption { get;set }
}

```

**Comment by stephentoub** [comment]
> Default = 0, // As stephentoub says, always return the same value regardless of machine
> Current, // Obey the current machine endian

What is the difference between these two?

**Comment by LeaFrock** [comment]
Well, I'm not sure if my understanding is right. Suppose that machine A uses BE as default, when machine B uses LE. 

`Default` will obey the rule of API itself which ignores all the differences between machines. A and B always produce the same outputs.
`Current` obey the rule of machine system so that the outputs are different from A and B.

**Comment by stephentoub** [comment]
> Current obey the rule of machine system so that the outputs are different from A and B.

As long as an algorithm isn't reinterpreting the raw bytes that make up an integer, endianness has no effect on the math involved, e.g. 2*3 is 6 regardless of whether you're on a little-endian or big-endian machine, 42 >> 3 is 5 regardless of whether you're on a little-endian or big-endian machine, (byte)0x12345678 is 0x78 regardless of whether you're on a little-endian or big-endian machine, etc., but `int i = 0x12345678; *(byte*)&i` _will_ differ, as it's bypassing the normal language rules for how to interpret the bytes and instead accessing a specific memory location directly, and thus will be 0x78 on a little-endian machine and 0x12 on a big-endian machine.

All of these algorithms just operate on the numbers, not the raw bytes, or if they do operate on the raw bytes, the algorithms themselves account for endianness by dictating the ordering in which the bytes need to be interpreted.  For example, the Crc32 algorithm's main loop is just:
https://github.com/dotnet/runtime/blob/68935248bcf6c555335336228ef42d3ccff4b277/src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs#L160-L165
There's nothing here endianness-specific: on both little-endian and big-endian machines, it'll produce the exact same uint value (the underlying memory will store the bytes that make up that uint differently, but the uint will have the same value).  Or in XXH64, for example, if bytes need to be read and re-interpreted as an integer, it's done so with a fixed endianness regardless of the machine's architecture:
https://github.com/dotnet/runtime/blob/68935248bcf6c555335336228ef42d3ccff4b277/src/libraries/System.IO.Hashing/src/System/IO/Hashing/XxHash64.State.cs#L81
and that's dictated by the algorithm, e.g. https://github.com/Cyan4973/xxHash/blob/f9155bd4c57e2270a4ffbb176485e5d713de1c9b/doc/xxhash_spec.md states "However, a given variant shall produce exactly the same output, irrespective of the cpu / os used. In particular, the result remains identical whatever the endianness and width of the cpu is."

In other words, regardless of whatever additional APIs .NET wants to layer on top, these algorithms produce a number and that number is computed to have a value that's the same regardless of endianness.  The only reason endianness comes into play in the discussions about these NonCryptographicHashAlgorithm APIs is _because_ they then transform the resulting numbers into bytes, as the {Try}Hash methods all only return the data as `byte[]` or written into a `Span<byte>`.  But when talking about the actual numerical result of the algorithm, endianness shouldn't be a factor, which is why I don't think anything to do with endianness should be a part of the APIs being proposed here that are just returning int/long.

**Comment by LeaFrock** [comment]
> As long as an algorithm isn't reinterpreting the raw bytes that make up an integer, endianness has no effect on the math involved
>...
>All of these algorithms just operate on the numbers, not the raw bytes, or if they do operate on the raw bytes, the algorithms themselves account for endianness by dictating the ordering in which the bytes need to be interpreted

Got it. Thanks for your explanation and patience 😄!

The last worry I have is that if a library has already used `(Try)Hash` and `BitConverter` with `byte[]` and appointed endianness to get a `int`/`long` value, it would meet potential break changes after switching to the new API. But it's not a big problem as more detailed docs will help that.

>Almost all of the examples I've found are folks who want the values as integers, and thus are having to reconstruct that from the bytes we spent energy producing

What I regret is that the proposed APIs and related discussions did not occur before .NET 6 (adding `System.IO.Hashing`).

**Comment by jods4** [comment]
> For digest algorithms (a.k.a. hash algorithms) [...] we want the resulting digest bytes to be the source of truth. **Always.**
> So if we expose `GetHashCodeAsXxx` APIs on these types, it really should follow two constraints.

Amusingly, one of the most fundamental method of .net is `Object.GetHashCode()` and it returns an `int` ;)

It has the same name as `GetHashCodeAsXxx` and it fits the terminology of "a variable number of bytes in" and "a fixed-length sequence of bytes out.", especially in cases like `string` or even `byte[]`!

Maybe it's a huge mistake in framework design, but I have a hunch that if we designed .net from scratch today we wouldn't make it `byte[] Object.GetHashCode()`.

**Comment by GrabYourPitchforks** [comment]
@jods4 - `object.GetHashCode` is very specifically intended for usage within a dictionary and other keyed hash collections in-proc. It is not meant to be persisted. It does not need to be stable across architectures, or even across back-to-back executions of the app on the same machine! Its underlying algorithm is strictly an implementation detail and callers cannot rely on it to be a member of any specific algorithm family.

@stephentoub - since `Int128` is net7+ only, do we need to condition the friendly one-shot helper on Xxhash128 to be exposed only in net7+?

**Comment by stephentoub** [comment]
> since Int128 is net7+ only, do we need to condition the friendly one-shot helper on Xxhash128 to be exposed only in net7+?

Yes, it'll be .NET 7 only.

**Comment by terrajobst** [comment]
> That our APIs force you to either allocate an array or stackalloc some space into which to store the result, figure out that it's been written as a big-endian value, and then convert it back to being a number is a pit of failure.

LOL. I literally just wrote that code yesterday.

**Comment by bartonjs** [comment]
Note that the XxHash128-Int128 members won't be available in netstandard2.0.

```C#
namespace System.IO.Hashing;

public partial sealed class Crc32 : NonCryptographicHashAlgorithm
{
    public static uint HashToUInt32(ReadOnlySpan<byte> source);
    public uint GetCurrentHashAsUInt32();
}

public partial sealed class XxHash32 : NonCryptographicHashAlgorithm
{
    public static uint HashToUInt32(ReadOnlySpan<byte> source, int seed = 0);
    public uint GetCurrentHashAsUInt32();
}

public partial sealed class Crc64 : NonCryptographicHashAlgorithm
{
    public static ulong HashToUInt64(ReadOnlySpan<byte> source);
    public ulong GetCurrentHashAsInt64();
}

public partial sealed class XxHash64 : NonCryptographicHashAlgorithm
{
    public static ulong HashToUInt64(ReadOnlySpan<byte> source, long seed = 0);
    public ulong GetCurrentHashAsUInt64();
}

public partial sealed class XxHash3 : NonCryptographicHashAlgorithm
{
    public static ulong HashToUInt64(ReadOnlySpan<byte> source, long seed = 0);
    public ulong GetCurrentHashAsUInt64();
}

public partial sealed class XxHash128 : NonCryptographicHashAlgorithm
{
    public static Int128 HashToUInt128(ReadOnlySpan<byte> source, long seed = 0);
    public Int128 GetCurrentHashAsUInt128();
}
```

**Comment by teo-tsirpanis** [comment]
[The entry in `dotnet/apireviews`](https://github.com/dotnet/apireviews/tree/main/2022/11-08-quick-reviews#xxhash128) got a bit messed up with #69590.

**Comment by xoofx** [comment]
Just to make sure, for `XxHash128`, it is actually a `UInt128` that we should return, right?

```diff
public partial sealed class XxHash128 : NonCryptographicHashAlgorithm
{
-    public static Int128 HashToUInt128(ReadOnlySpan<byte> source, long seed = 0);
+    public static UInt128 HashToUInt128(ReadOnlySpan<byte> source, long seed = 0);
-    public Int128 GetCurrentHashAsUInt128();
+    public UInt128 GetCurrentHashAsUInt128();
}
```

**Comment by stephentoub** [comment]
yup, UInt128, thanks
